File: CSX/Callbacks & Higher-order Functions/eitherFilter.js
//EITHERFILTER

//expected input(s): array, callback function 1, callback function 2
//expected output(s): new array
/*expected result(s): pass each element from array as argument for both callback functions;
if element satisfies either function, push it to a new array, then return the array.
*/

function eitherFilter(array, callback1, callback2) {
  //declare the output array:
  let outputArray = [];

  //declare a function to push truthy input array elements into the output array:
  function pusher(element) {
    if (callback1(element) || callback2(element)) outputArray.push(element);
  }

  //run the pusher function using each element of input array as argument:
  array.forEach(pusher);

  //return the new array:
  return outputArray;


}

const arrOfNums = [10, 35, 105, 9];
const integerSquareRoot = n => Math.sqrt(n) % 1 === 0;
const over100 = n => n > 100;

//eitherFilter(array, callback function 1, callback function 2)
/*"For each element in the array [10, 35, 105, 9], pass it as an argument to both callback functions.
If either one of the functions returns 'true', then push that element to a new array.
Return that array." */
console.log(eitherFilter(arrOfNums, integerSquareRoot, over100));

File: CSX - Python/Callbacks & Higher-order Functions/either_filter.py
from math import sqrt

def either_filter(list, cb1, cb2):
    return [x for x in list if cb1(x) or cb2(x)]

list_of_nums = [10, 35, 105, 9]

def integer_square_root(n):
    return sqrt(n) % 1 == 0

def over_100(n):
    return n > 100

print(either_filter(list_of_nums, integer_square_root, over_100))
